pyscf/lib/dft/libxc_itrf.c

```diff
@@ -362,6 +363,124 @@ static void _eval_xc(xc_func_type *func_x, int spin, int np,
         }
 }
 
+typedef struct {
+        int id;
+        double *params;
+        size_t size;
+} ext_params;
+
+static ext_params *custom_ext_params = NULL;
+static int custom_ext_params_count = 0;
+
+int set_ext_params(int id, double *params, size_t size)
+{
+        // Update existing entry if exists
+        for (int i = 0; i < custom_ext_params_count; i++) {
+                ext_params *p = &custom_ext_params[i];
+                if (p->id == id) {
+                        if (p->size != size) {
+                                fprintf(stderr, "Inconsistent parameter size. Expected %lu but got %lu\n", p->size, size);
+                                raise_error -1;
+                        }
+                        memcpy(p->params, params, size * sizeof(double));
+                        return 0;
+                }
+        }
+
+        {
+                // Check params size
+                xc_func_type func;
+                if (xc_func_init(&func, id, XC_UNPOLARIZED) != 0){
+                        fprintf(stderr, "XC functional %d not found\n", id);
+                        raise_error -1;
+                }
+                int ref_size = func.info->ext_params.n;
+                xc_func_end(&func);
+                if (ref_size != size) {
+                        fprintf(stderr, "Inconsistent parameter size. Expected %d but got %lu.\n", ref_size, size);
+                        raise_error -1;
+                }
+        }
+
+        // Create new entry
+        int i = custom_ext_params_count++;
+        custom_ext_params = realloc(custom_ext_params, custom_ext_params_count * sizeof(ext_params));
+        ext_params *p = &custom_ext_params[i];
+        p->id = id;
+        p->params = malloc(size * sizeof(double));
+        memcpy(p->params, params, size * sizeof(double));
+        p->size = size;
+        return 0;
+}
+
+int remove_ext_params(int id)
+{
+        for (int i = 0; i < custom_ext_params_count; i++) {
+                if (custom_ext_params[i].id == id) {
+                        free(custom_ext_params[i].params);
+                        for (int j = i; j < custom_ext_params_count - 1; j++) {
+                                custom_ext_params[j] = custom_ext_params[j + 1];
+                        }
+                        custom_ext_params_count--;
+                        return 0;
```

Comment:
The deallocater isn't working properly. You also need to reallocate the cache size.

Reply:
Fixed